### PR TITLE
adds an apply function and cleans adj_jac_apply to use it

### PR DESCRIPTION
**Pull request by SteveBronder**
## Summary

1. An apply function on tuples which is a backwards import from C++17's std::apply
2. Fixes to broadcast_array so that the scalar implimentation has operator's [ ] and = and a few others defined
3. adds the inline keyword to adj_jac_apply methods and uses eigen map's instead of loops.

## Tests

Tests can be run with
```
./runTests.py test/unit/math/prim/functor/apply_test.cpp \
  test/unit/math/rev/functor/adj_jac_apply_test \
```

## Side Effects

None

## Release notes

Added an apply function that backports C++17's `apply()` over tuples. Simple cleanups for `broadcast_array` and `adj_jac_apply`

## Checklist

- [x] Math issue #1616

- [x] Copyright holder: Steve Bronder, Ben Bales, Sebasitan Weber

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/prim/meta/broadcast_array.hpp [hunk 1]
```diff
@@ -30,22 +32,34 @@ class broadcast_array {
   void operator=(const Y& m) {
     prim_ = m[0];
   }
+  explicit broadcast_array(const broadcast_array<T>& other) = default;
```
Comment (bbbales2) [inline stan/math/prim/meta/broadcast_array.hpp]: If these are default, why do they need defined?

File: stan/math/rev/functor/adj_jac_apply.hpp [hunk 2]
```diff
@@ -272,53 +239,55 @@ struct adj_jac_vari : public vari {
...
   template <int R, int C, typename... Pargs>
-  void prepare_x_vis(const Eigen::Matrix<var, R, C>& x, const Pargs&... args) {
+  inline void prepare_x_vis(const Eigen::Matrix<var, R, C>& x,
+                            const Pargs&... args) {
     static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    for (int i = 0; i < x.size(); ++i) {
-      x_vis_[offsets_[t] + i] = x(i).vi_;
-    }
+    using write_map = Eigen::Map<Eigen::Matrix<vari*, R, C>>;
+    write_map(this->x_vis_ + this->offsets_[t], x.rows(), x.cols()) = x.vi();
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/functor/adj_jac_apply.hpp]: I was interested in testing out whether the eigen stuff would be better and wrote a benchmark for it. TL;DR it seems to be for smaller matrices but for large ones just doing the loop is faster

```
git clone --recursive https://github.com/seantalts/perf-math.git
git checkout eigen_pf_var
make eigen_eq_add_var.o && make eigen_eq_add_var
./eigen_pf --benchmark_repetitions=20 --benchmark_report_aggregates_only=true
```
```
--------------------------------------------------------------------------------
Benchmark                                      Time             CPU   Iterations
--------------------------------------------------------------------------------
eq_eig_bench/10/manual_time_mean            45.7 ns         2502 ns           20
eq_eig_bench/10/manual_time_median          45.1 ns         2484 ns           20
eq_eig_bench/10/manual_time_stddev          1.61 ns         78.7 ns           20
eq_eig_bench/110/manual_time_mean           5198 ns       382708 ns           20
eq_eig_bench/110/manual_time_median         3442 ns       326092 ns           20
eq_eig_bench/110/manual_time_stddev         3497 ns       108755 ns           20
eq_eig_bench/210/manual_time_mean         111945 ns      3412061 ns           20
eq_eig_bench/210/manual_time_median       113402 ns      3489592 ns           20
eq_eig_bench/210/manual_time_stddev        16526 ns       467508 ns           20
eq_loop_bench/10/manual_time_mean           67.8 ns         3553 ns           20
eq_loop_bench/10/manual_time_median         65.0 ns         3484 ns           20
eq_loop_bench/10/manual_time_stddev         9.56 ns          341 ns           20
eq_loop_bench/110/manual_time_mean          9090 ns       514242 ns           20
eq_loop_bench/110/manual_time_median        8842 ns       517479 ns           20
eq_loop_bench/110/manual_time_stddev        1358 ns        37058 ns           20
eq_loop_bench/210/manual_time_mean         98055 ns      2949144 ns           20
eq_loop_bench/210/manual_time_median       94485 ns      2824535 ns           20
eq_loop_bench/210/manual_time_stddev       14113 ns       500216 ns           20
add_eig_bench/10/manual_time_mean           47.6 ns         1011 ns           20
add_eig_bench/10/manual_time_median         49.5 ns         1027 ns           20
add_eig_bench/10/manual_time_stddev         8.00 ns          150 ns           20
add_eig_bench/110/manual_time_mean          3938 ns       120977 ns           20
add_eig_bench/110/manual_time_median        3993 ns       119648 ns           20
add_eig_bench/110/manual_time_stddev         701 ns        19583 ns           20
add_eig_bench/210/manual_time_mean         21713 ns       469973 ns           20
add_eig_bench/210/manual_time_median       19733 ns       478979 ns           20
add_eig_bench/210/manual_time_stddev        5331 ns        42877 ns           20
add_loop_bench/10/manual_time_mean          71.9 ns         1139 ns           20
add_loop_bench/10/manual_time_median        67.0 ns         1113 ns           20
add_loop_bench/10/manual_time_stddev        13.8 ns          118 ns           20
add_loop_bench/110/manual_time_mean         4071 ns       127772 ns           20
add_loop_bench/110/manual_time_median       3999 ns       127692 ns           20
add_loop_bench/110/manual_time_stddev        525 ns         5962 ns           20
add_loop_bench/210/manual_time_mean        21797 ns       454361 ns           20
add_loop_bench/210/manual_time_median      19978 ns       463360 ns           20
add_loop_bench/210/manual_time_stddev       5991 ns        26966 ns           20
```

I liked the eigen style but at best the eigen style is the same within a margin of error as the loop so if it's more readable for you to have the loop then I'm fine with just a loop here. We should also switch this in reduce sum

File: stan/math/rev/functor/adj_jac_apply.hpp [hunk 3]
```diff
@@ -188,70 +154,71 @@ struct adj_jac_vari : public vari {
...
     offsets_[t] = count;
     return count_memory(count, args...);
   }
 
   template <typename... Pargs>
-  size_t count_memory(size_t count, const var& x, const Pargs&... args) {
+  inline size_t count_memory(size_t count, const var& x, const Pargs&... args) {
     static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
+    this->offsets_[t] = count;
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/functor/adj_jac_apply.hpp]: I'm fine with removing these, though personally I find when reading large classes using `this` helps remove ambiguity. When I start reading the line I already know we are manipulating a class member

File: stan/math/rev/functor/adj_jac_apply.hpp [hunk 4]
```diff
@@ -71,10 +37,10 @@ void build_y_adj(vari** y_vi, const std::array<int, size>& M, double& y_adj) {
  * @param[out] y_adj reference to std::vector where adjoints are to be stored
  */
 template <size_t size>
-void build_y_adj(vari** y_vi, const std::array<int, size>& M,
-                 std::vector<double>& y_adj) {
-  y_adj.resize(M[0]);
-  for (size_t m = 0; m < y_adj.size(); ++m) {
+inline void build_y_adj(vari** y_vi, const std::array<int, size>& M,
+                        std::vector<double>& y_adj) {
+  y_adj.reserve(M[0]);
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/functor/adj_jac_apply.hpp]: I'm fine with changing it, but I'm not sure if I'm following what you are saying here.

> Reserve doesn't change the size of the vector. Either reserve + push back or resize.

Yes reserve only changes the capacity while leaving new reserved memory uninitialized. Resize will shrink to fit or allocate new memory with default initialization.

> The M[0] change is a whatever, but if the compiler can't be trusted to optimize out y_adj.size() we're done anyway.

I'm not sure what you mean by optimize out y_adj.size().  We bring in `y_adj` by reference so idt the compiler would have an easy time at all deducing whether size() changes over iterations. 

> y_adj.size() makes the conditions on the loop be written in terms of the things in the loop as well, but that's also a minor point (though this should have helped the tests catch the reserve problem).

Running benchmark reserve seems faster here by a good bit

```
-------------------------------------------------------------
Benchmark                   Time             CPU   Iterations
-------------------------------------------------------------
resize_bench/10          35.3 ns         33.6 ns     18482503
resize_bench/1010         472 ns          470 ns      1634294
resize_bench/2010        1003 ns          917 ns       975563
reserve_bench/10         24.8 ns         24.8 ns     26297762
reserve_bench/1010        273 ns          264 ns      3172617
reserve_bench/2010        514 ns          514 ns      1760548
```

I'm not sure if I'm following what the problem is with reserve here. You can write up to the capacity of the vector no? (which we know ahead of the loop is M[0])

[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/rev/functor/adj_jac_apply.hpp]: Reserve doesn't change the size of the vector. It still thinks it's a size zero (so if you checked y_adj.size(), it'd say 0 still). Reserve just makes sure you can do push_backs without having to do any memory reallocation.

Doing reserve + push_back was faster here: https://github.com/stan-dev/math/blob/5f64b49f29995cd2e53674d1c000d09cc8c7d40b/stan/math/rev/functor/reduce_sum.hpp#L99 at least, so reserve being fast kinda makes sense there. I'm not totally sure why but it seems it is so.

[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/functor/adj_jac_apply.hpp]: > Reserve doesn't change the size of the vector. 

Yep! That's what I meant by this

>> Yes reserve only changes the capacity while leaving new reserved memory uninitialized. 

That's why the loop goes over `M[0]`. I'm not following what's wrong with this?

> Doing reserve + push_back was faster here:

In the same repo as the other performance tests below I have a microbenchmark for this `resize_vs_reserve.cpp`. The below is iterating over vectors of size 10, 1010, and 2010. and doing a resize/reserve and assignment with `operator[]` whle `reserve_pushback` does a reserve with a push_back.

```
----------------------------------------------------------------------
Benchmark                            Time             CPU   Iterations
----------------------------------------------------------------------
resize_bench/10                   38.0 ns         35.8 ns     18837778
resize_bench/1010                  761 ns          723 ns      1261559
resize_bench/2010                 1462 ns         1421 ns       475598
reserve_bench/10                  24.9 ns         24.1 ns     33730549
reserve_bench/1010                 335 ns          303 ns      2143224
reserve_bench/2010                 530 ns          494 ns      1407691
reserve_pushback_bench/10         32.6 ns         29.0 ns     24350310
reserve_pushback_bench/1010        865 ns          819 ns       918158
reserve_pushback_bench/2010       2243 ns         2212 ns       540014
```

Doing a push_back forces it to check if it has enough space, just doing `[i]` does not.

[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/rev/functor/adj_jac_apply.hpp]: I don't think it's valid to use the ```std::vector``` like that though. Reserve is the prevent reallocations. It's either resize + operator[] or reserve + push_back, not reserve + operator[]. Someone in this thread calls it undefined behavior: https://stackoverflow.com/questions/13029299/stdvectorresize-vs-stdvectorreserve .

[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/rev/functor/adj_jac_apply.hpp]: But that is interesting that they all perform so differently. I'd have expected everything there to be basically the same.

File: stan/math/prim/meta/broadcast_array.hpp [hunk 5]
```diff
@@ -13,14 +13,16 @@ namespace internal {
 template <typename T>
 class broadcast_array {
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/prim/meta/broadcast_array.hpp]: These were in the reduce_sum pr so I added them here @wds15 did you make these changes / can I get rid of them?

File: stan/math/prim/meta/broadcast_array.hpp (hunk 1 above)
Comment (SteveBronder) [reply stan/math/prim/meta/broadcast_array.hpp]: For some reason in the  `reduce_sum` PR we did
```
broadcast_array& operator=(const broadcast_array& other) = default;
```

I'm not sure why and can delete these or (@wds15 did you put the `operator=` one here for a reason?

File: stan/math/rev/functor/adj_jac_apply.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/functor/adj_jac_apply.hpp]: Huh! Yeah I'll switch this back to resize then

https://godbolt.org/z/6_DRyT

File: stan/math/rev/functor/adj_jac_apply.hpp [hunk 6]
```diff
@@ -272,7 +239,8 @@ struct adj_jac_vari : public vari {
    * recursively)
    */
   template <int R, int C, typename... Pargs>
-  void prepare_x_vis(const Eigen::Matrix<var, R, C>& x, const Pargs&... args) {
+  inline void prepare_x_vis(const Eigen::Matrix<var, R, C>& x,
+                            const Pargs&... args) {
```
Comment (andrjohns) [inline stan/math/rev/functor/adj_jac_apply.hpp]: Would it be feasible in this pull to generalise the Eigen inputs to ```adj_jac_apply``` or should that be a separate PR?

Where this could be something like:
```
  template <typename T, typename... Pargs, require_eigen_st<is_var, T>...>
  void prepare_x_vis(const T& x, const Pargs&... args) {
    static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
    Eigen::Map<matrix_vi>(&x_vis_[offsets_[t]], x.rows(), x.cols()) = x.matrix().vi();
    prepare_x_vis(args...);
  }
```

File: stan/math/prim/functor/apply.hpp [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * TODO: replace this with implementation in C++ std when C++17 is available
+ *
+ * @tparam F Type of functor
+ * @tparam Tuple Type of tuple containing arguments
+ * @param f functor callable
+ * @param t tuple of arguments
+ */
+template <class F, class Tuple>
+constexpr decltype(auto) apply(F&& f, Tuple&& t) {
```
Comment (t4c1) [inline stan/math/prim/functor/apply.hpp]: We have a really similar function here: https://github.com/stan-dev/math/blob/develop/stan/math/prim/meta/index_apply.hpp
Do we need both?

File: stan/math/rev/functor/adj_jac_apply.hpp (hunk 6 above)
Comment (bbbales2) [reply stan/math/rev/functor/adj_jac_apply.hpp]: > Would it be feasible in this pull to generalise the Eigen inputs

Yeah, separate PR though. Would want to add testing and such. And ideally it should come with a reason for doing it (and function or whatnot we'd be autodiffing).

I'm also curious if there's any forwarding/whatnot we could do to make this match efficiency in situations like the vector ops you were looking at (https://discourse.mc-stan.org/t/adj-jac-apply/5163/6)

File: stan/math/prim/functor/apply.hpp (hunk 7 above)
Comment (bbbales2) [reply stan/math/prim/functor/apply.hpp]: Probly not. Steve maxed out the forwarding on this one and there are some tests, so I think we go this route.

Looks like index_apply is in the kernel_generator. Do you see any problems with using apply there? If not mind swapping it out, verifying the tests for kernel generator still pass, removing index_apply, and pushing the changes here?

File: stan/math/prim/meta/broadcast_array.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/meta/broadcast_array.hpp]: Huh, strange, I think just boot these changes if @wds15 doesn't recognize them. Maybe they got accidentally included from something else.

File: stan/math/rev/functor/adj_jac_apply.hpp (hunk 6 above)
Comment (andrjohns) [reply stan/math/rev/functor/adj_jac_apply.hpp]: Cool cool, I'll open an issue and tackle that one after this pull is finished

Comment (bbbales2) [reply stan/math/rev/functor/adj_jac_apply.hpp]: Feel free to think about other problems/opportunities with adj_jac_apply too. Seems like it never caught on like I hoped.

Maybe there are ways to make it simpler, or maybe it's a lack of documentation, I dunno.

File: stan/math/prim/functor/apply.hpp (hunk 7 above)
Comment (t4c1) [reply stan/math/prim/functor/apply.hpp]: I think that should work. Though I will have to undo part of refactoring in #1769. Actually I would prefer to do that in a separate PR after this and #1769 are merged - otherwise there will be a number of merge conflicts.

Comment (bbbales2) [reply stan/math/prim/functor/apply.hpp]: Sure, seems fine to me. Can note this issue somewhere and track it?

File: stan/math/prim/meta/broadcast_array.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (bbbales2) [reply stan/math/prim/meta/broadcast_array.hpp]: @SteveBronder I think the changes came from here: https://github.com/stan-dev/math/commit/c53d29d1d0365b17160517b40adf0c1b2d83596d#diff-1632e3c2d4d0d4ec6a9367d12f64dfcf

Any chance these are accidental includes from a different branch or something?

File: stan/math/rev/functor/adj_jac_apply.hpp [hunk 8]
```diff
@@ -188,70 +154,71 @@ struct adj_jac_vari : public vari {
...
-  size_t count_memory(size_t count, const int& x, const Pargs&... args) {
+  inline size_t count_memory(size_t count, const int& x, const Pargs&... args) {
     static constexpr int t = sizeof...(Targs) - sizeof...(Pargs) - 1;
-    offsets_[t] = count;
+    this->offsets_[t] = count;
     return count_memory(count, args...);
   }
 
-  size_t count_memory(size_t count) { return count; }
+  size_t count_memory(size_t count) const { return count; }
```
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/rev/functor/adj_jac_apply.hpp]: Yes but those ones actually change internal state while this call does not.

**Comment by SteveBronder** [comment]
@bbbales2 I removed the use of `this`, the changes to broadcast_array, and use `resize` again so I think this is ready for review!

File: stan/math/prim/meta/broadcast_array.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (SteveBronder) [reply stan/math/prim/meta/broadcast_array.hpp]: Possibly, I kicked them out

**Comment by SteveBronder** [comment]
I concur!

File: stan/math/rev/functor/adj_jac_apply.hpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply stan/math/rev/functor/adj_jac_apply.hpp]: We haven't used `this` anywhere else like this. 

We use a trailing underscore on all and only member variables to distinguish them visually in the code.

File: stan/math/rev/functor/adj_jac_apply.hpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (bob-carpenter) [reply stan/math/rev/functor/adj_jac_apply.hpp]: +1 for loop for readability.  I can read the Eigen, but find the loop much more readable.

File: stan/math/prim/meta/broadcast_array.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (wds15) [reply stan/math/prim/meta/broadcast_array.hpp]: I recall that I messed with this code a while ago to get things going with a different approach. If reduce_sum does not need this in its current design, then just ditch it.

Sorry, I went through a lot of things.

File: stan/math/prim/meta/broadcast_array.hpp (hunk 1 above)
Comment (wds15) [reply stan/math/prim/meta/broadcast_array.hpp]: Maybe. Likely not needed any longer.
